src/modules/nav/nav.service.ts

```diff
@@ -0,0 +1,103 @@
+import { Injectable } from '@angular/core';
+import { Router, NavigationStart } from '@angular/router';
+
+import { SkyAppConfig } from '@blackbaud/skyux-builder/runtime';
+
+import { StacheNavLink } from './nav-link';
+
+@Injectable()
+export class StacheNavService {
+  private activeRoutes: StacheNavLink[];
+
+  public constructor(
+    private appConfig: SkyAppConfig,
+    private router: Router
+  ) {
+    router.events.subscribe((val: any) => {
+      if (val instanceof NavigationStart) {
+        this.clearActiveRoutes();
+      }
+    });
+  }
+
+  public getActiveUrl(): string {
+    return this.router.url.split('#')[0];
+  }
+
+  public getActiveRoutes(): StacheNavLink[] {
+    if (this.activeRoutes) {
+      return this.clone(this.activeRoutes) as StacheNavLink[];
+    }
+
+    let rootPath = this.getActiveUrl().replace(/^\//, '').split('/')[0];
+    let routes = this.clone(this.appConfig.runtime.routes);
+
+    routes = routes.filter((route: any) => route.routePath.indexOf(rootPath) === 0);
+
+    routes = routes.map((route: any) => {
+      return {
+        segments: route.routePath.split('/'),
+        path: route.routePath
+      };
+    });
+
+    routes = [{
+      path: rootPath,
+      segments: [rootPath],
+      children: this.orderRoutes(routes, rootPath)
+    }];
+
+    this.activeRoutes = this.formatRoutes(routes);
+
+    return this.clone(this.activeRoutes) as StacheNavLink[];
+  }
+
+  public clearActiveRoutes() {
+    this.activeRoutes = undefined;
+  }
+
+  private orderRoutes(routes: any[], parentPath: string): any[] {
+    const ordered: any[] = [];
+    const depth = parentPath.split('/').length + 1;
+
+    routes.forEach(route => {
+      const routeDepth = route.segments.length;
+      const isChildRoute = (depth === routeDepth && route.path.indexOf(parentPath) > -1);
+
+      if (isChildRoute) {
+        route.children = this.orderRoutes(routes, route.path);
+        ordered.push(route);
+      }
+    });
+
+    return ordered;
+  }
+
+  private formatRoutes(routes: any[]): StacheNavLink[] {
+    let formatted = routes.map(route => {
+      return {
+        name: this.getNameFromPath(route.segments[route.segments.length - 1]),
+        path: `/${route.path}`,
+        children: this.formatRoutes(route.children)
+      } as any;
+    });
+
+    return formatted as StacheNavLink[];
+  }
+
+  private getNameFromPath(path: string): string {
+    path = path.replace(/-/g, ' ');
+    return this.toTitleCase(path);
+  }
+
+  private toTitleCase(phrase: string): string {
+    return phrase
+      .split(' ')
+      .map(word => word.charAt(0).toUpperCase() + word.slice(1))
+      .join(' ');
+  }
+
+  private clone(thing: any): any {
```

Comment:
In this case, the method is simply cloning a thing (could be a string, an array, or whatever).